ml/rl/workflow/data_fetcher.py

```diff
@@ -0,0 +1,274 @@
+)
+from pyspark.sql.functions import col, udf, crc32, asc
+from ml.rl.workflow.types import TableSpec, ParquetDataset
+from ml.rl.workflow.utils import get_spark_session
+import logging
+
+logger = logging.getLogger(__name__)
+
+# for normalizing crc32 output
+MAX_UINT32 = 4294967295
+
+
+def calc_custom_reward(df, custom_reward_expr: str):
+    # get current session
+    spark = get_spark_session()
+    temp_table_name = "_tmp_calc_reward_df"
+    temp_reward_name = "_tmp_reward_col"
+    df.registerTempTable(temp_table_name)
+    df = spark.sql(
+        f"SELECT *, CAST(COALESCE({custom_reward_expr}, 0) AS FLOAT) as {temp_reward_name} FROM {temp_table_name}"
+    )
+    return df.drop("reward").withColumnRenamed(temp_reward_name, "reward")
+
+
+def calc_reward_multi_step(df, multi_step: int, gamma: float):
+    def multi_step_reward(reward_arr):
+        assert (
+            len(reward_arr) <= multi_step
+        ), f"reward arr has length {len(reward_arr)} > {multi_step}"
+        result = 0.0
+        for i, reward in enumerate(reward_arr):
+            result += reward * (gamma ** i)
+        return result
+
+    multi_step_reward_udf = udf(multi_step_reward, FloatType())
+    return df.withColumn("reward", multi_step_reward_udf("reward"))
+
+
+def perform_preprocessing(
+    df,
+    state_keys: List[int],
+    actions: List[str],
+    metrics_keys: List[str],
+    use_seq_num_diff_as_time_diff: bool,
+    sort: bool = False,
+    custom_reward_expr: Optional[str] = None,
+    sample_range: Optional[Tuple[float, float]] = None,
+    multi_step: Optional[int] = None,
+    gamma: Optional[float] = None,
+):
+    """ Perform preprocessing of given dataframe df.
+    Preprocessing steps include calculating the reward, 
+    performing sparse-to-dense for mapped columns like state_features
+    and metrics, and subsampling based on sample_range. 
+    If multi_step is set (with gamma), then we assume multi_step RL setting.
+    """
+    if sample_range:
+        assert (
+            0.0 <= sample_range[0]
+            and sample_range[0] <= sample_range[1]
+            and sample_range[1] <= 100.0
+        ), f"{sample_range} is invalid."
+    assert (
+        not use_seq_num_diff_as_time_diff or not multi_step
+    ), "use_seq_num_diff_as_time_diff and multi_step should not work together"
+
+    # after this, reward column should be set to be the reward now
+    if custom_reward_expr is not None:
+        df = calc_custom_reward(df, custom_reward_expr)
+    elif multi_step is not None:
+        assert gamma is not None
+        df = calc_reward_multi_step(df, multi_step, gamma)
+    # assume single step case reward is already a column
+
+    def get_step(next_col):
+        """ get step count """
+        if multi_step is not None:
+            return min(len(next_col), multi_step)
+        else:
+            return 1
+
+    get_step_udf = udf(get_step, LongType())
+    df = df.withColumn("step", get_step_udf("next_state_features"))
```

Comment:
I'm not sure if anything changed, but someone on the internet said that `withColumn` can be slow and they recommended `df.select()` instead.

https://medium.com/@manuzhang/the-hidden-cost-of-spark-withcolumn-8ffea517c015

Reply:
I'm not sure about this either. 
If I use select, I would need to select the whole dataframe except that one column. I think syntactically withColumn is still clearer, and since we're only doing it ~10 times instead of thousands (like the post), it's probably not a big deal.